### PR TITLE
Feat/cooking recipe tr

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

**Review by ChengshuLi — APPROVED** [review]

File: omnigibson/transition_rules.py [hunk 1]
```diff
@@ -1168,43 +1312,292 @@ def _generate_conditions(cls):
...
+    Transition mixing rule that approximates cooking recipes via a container and heatsource
+    """
+    # Counter that increments monotonically
+    cls.COUNTER = 0
+
+    # Maps recipe name to current number of consecutive heating steps
+    cls._HEAT_STEPS = None
+
+    # Maps recipe name to the last timestep that it was active
+    cls._LAST_HEAT_TIMESTEP = None
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/transition_rules.py:1375]: Nvm, I think your method works pretty cleanly.

File: omnigibson/transition_rules.py [hunk 2]
```diff
@@ -1113,12 +1222,37 @@ def _execute_recipe(cls, container, recipe, in_volume):
...
+
+        # Spawn in new objects
+        for category, n_instances in recipe["output_objects"].items():
+            n_category_objs = len(og.sim.scene.object_registry("category", category, []))
+            models = get_all_object_category_models(category=category)
+            for i in range(n_instances):
+                obj = DatasetObject(
+                    name=f"{category}_{n_category_objs + i}",
+                    category=category,
+                    model=model,
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/transition_rules.py:1248]: fixed

File: omnigibson/transition_rules.py [hunk 3]
```diff
@@ -1113,12 +1222,37 @@ def _execute_recipe(cls, container, recipe, in_volume):
...
         volume += sum(obj.volume for obj in objs_to_remove)
 
+        # Define callback for spawning new objects inside container
+        def _spawn_object_in_container(obj):
+            # We will either sample Inside or OnTop, based on the relative AABB extents of the container link
+            # vs. the object link
+            container_aabb = container.states[ContainedParticles].link.aabb_extent
+            obj_aabb = obj.aabb_extent
+            state = Inside if np.all(container_aabb > obj_aabb) else OnTop
+            obj.states[state].set_value(container, True)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/transition_rules.py:1238]: added assert

File: omnigibson/transition_rules.py [hunk 4]
```diff
@@ -873,22 +878,31 @@ def __init_subclass__(cls, **kwargs):
...
         """
         # Store information for this recipe
         cls._RECIPES[name] = {
-            "input_objects": input_objects,
-            "input_systems": input_systems,
-            "output_system": output_system,
+            "name": name,
+            "input_objects": dict() if input_objects is None else input_objects,
+            "input_systems": [] if input_systems is None else input_systems,
+            "output_objects": dict() if output_objects is None else output_objects,
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/transition_rules.py:908]: added assert to prevent both

**Pull request merged**